go/mysql/binlog_event_compression.go

```diff
@@ -41,35 +44,63 @@ const (
 	payloadUncompressedSizeField
 )
 
-// Compression algorithms that are supported (only zstd today
-// in MySQL 8.0):
-// https://dev.mysql.com/doc/refman/8.0/en/binary-log-transaction-compression.html
 const (
+	// Compression algorithms that are supported (only zstd today
+	// in MySQL 8.0):
+	// https://dev.mysql.com/doc/refman/8.0/en/binary-log-transaction-compression.html
 	TransactionPayloadCompressionZstd = 0
 	TransactionPayloadCompressionNone = 255
-)
 
-var TransactionPayloadCompressionTypes = map[uint64]string{
-	TransactionPayloadCompressionZstd: "ZSTD",
-	TransactionPayloadCompressionNone: "NONE",
-}
+	// Length of the binlog event header in the transaction payload.
+	headerLen = BinlogEventLenOffset + 4
 
-// Create a reader that caches decompressors. This is used for
-// smaller events that we want to handle entirely using in-memory
-// buffers.
-var zstdDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
+	// At what size should we switch from the in-memory buffer
+	// decoding to streaming mode which is much slower, but does
+	// not require everything be done in memory.
+	zstdInMemoryDecompressorMaxSize = 128 << (10 * 2) // 128MiB
+)
 
-// At what size should we switch from the in-memory buffer
-// decoding to streaming mode -- which is slower, but does not
-// require everything be done in memory.
-const zstdInMemoryDecompressorMaxSize = 128 << (10 * 2) // 128MiB
+var (
+	TransactionPayloadCompressionTypes = map[uint64]string{
+		TransactionPayloadCompressionZstd: "ZSTD",
+		TransactionPayloadCompressionNone: "NONE",
+	}
+
+	// Metrics.
+	compressedTrxPayloadsInMem       = stats.NewCounter("CompressedTransactionPayloadsInMemory", "The number of compressed binlog transaction payloads that were processed in memory")
+	compressedTrxPayloadsUsingStream = stats.NewCounter("CompressedTransactionPayloadsViaStream", "The number of compressed binlog transaction payloads that were processed using a stream")
+
+	// Create a concurrent stateless decoder that caches decompressors.
+	// This is used for smaller events that we want to handle entirely
+	// using in-memory buffers via DecodeAll.
+	statelessDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
```

Comment:
I know we check if `statelessDecoder` is `nil` later on, but is there a reason why we don't assign an `err` value and print that out when not-nil? It would give more information about _why_ `statelessDecoder` might be `nil`.

Reply:
Good idea. I moved the assignment to `init()` so that we can capture and log the error in: https://github.com/vitessio/vitess/commit/d5cbf05d3b0269a36a225b8b13b1952038c7f689